src/Tasks/SystemState.cs

```diff
@@ -548,12 +559,91 @@ out fileState.frameworkName
         }
 
         /// <summary>
-        /// Cached implementation of GetDirectories.
+        /// Reads in cached data from stateFiles to build an initial cache. Avoids logging warnings or errors.
         /// </summary>
-        /// <param name="path"></param>
-        /// <param name="pattern"></param>
-        /// <returns></returns>
-        private string[] GetDirectories(string path, string pattern)
+        internal static SystemState DeserializePrecomputedCaches(string[] stateFiles, TaskLoggingHelper log, Type requiredReturnType, GetLastWriteTime getLastWriteTime, AssemblyTableInfo[] installedAssemblyTableInfo)
+        {
+            SystemState retVal = new SystemState();
+            retVal.SetGetLastWriteTime(getLastWriteTime);
+            retVal.SetInstalledAssemblyInformation(installedAssemblyTableInfo);
+            retVal.isDirty = stateFiles.Length > 0;
+            HashSet<string> assembliesFound = new HashSet<string>();
+
+            foreach (string stateFile in stateFiles)
+            {
+                // Verify that it's a real stateFile; log message but do not error if not
+                SystemState sysState = (SystemState)DeserializeCache(stateFile, log, requiredReturnType, false);
+                if (sysState == null)
+                {
+                    continue;
+                }
+                foreach (DictionaryEntry kvp in sysState.instanceLocalFileStateCache)
```

Comment:
nit: make instanceLocalFileStateCache generic.

Reply:
I don't understand. Do you mean to make the type of instanceLocalFileStateCache more generic, change DictionaryEntry to var, make it so this can pull from something other than a system state by making the input to this loop instanceLocalFileStateCache instead of sysState, or something else?

Reply:
Make it a generic dictionary.